cpu/sam0_common/periph/freqm.c

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (C) 2023 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_sam0_common
+ * @ingroup     drivers_periph_freqm
+ * @{
+ *
+ * @file
+ * @brief       Frequency meter driver implementation
+ *
+ * @author      Urs Gompper <urs.gompper@ml-pa.com>
+ *
+ * @}
+ */
+
+#include "periph/freqm.h"
+
+static void _freqm_clock_init(uint8_t msr_clk_src, uint8_t ref_clk_src)
+{
+    /* Selection of the Generator and write Lock for FREQM_MSR */
+    GCLK->PCHCTRL[5].reg = GCLK_PCHCTRL_GEN(msr_clk_src) | GCLK_PCHCTRL_CHEN_Msk;
+    /* Wait for synchronization */
+    while ((GCLK->PCHCTRL[5].reg & GCLK_PCHCTRL_CHEN_Msk) != GCLK_PCHCTRL_CHEN_Msk) {}
+
+    /* Selection of the Generator and write Lock for FREQM_REF */
+    GCLK->PCHCTRL[6].reg = GCLK_PCHCTRL_GEN(ref_clk_src) | GCLK_PCHCTRL_CHEN_Msk;
+    /* Wait for synchronization */
+    while ((GCLK->PCHCTRL[6].reg & GCLK_PCHCTRL_CHEN_Msk) != GCLK_PCHCTRL_CHEN_Msk) {}
```

Comment:
You should use vendor files defines like `FREQM_GCLK_ID_MSR`

Reply:
> You should use vendor files defines like `FREQM_GCLK_ID_MSR`

This will work for `FREQM_GCLK_ID_MSR`, but unfortuenately `FREQM_GCLK_ID_REF` is not defined for samd51 and same54 devices, only for saml10 and saml11. They are defined in a newer version of the DFP as mentioned in https://github.com/RIOT-OS/RIOT/pull/20020#discussion_r1374057409

Reply:
Then it's fine to move these defines outside the driver scope as they are hardware-related.
@benpicco I remember vaguely a discussion about updating SAM0 vendor files a few months ago but there was an issue. Do you remember what was the issue ?

Reply:
> Do you remember what was the issue?

Microchip decided to rename all the structs and members, causing a lot of busywork if we ever want to support a new device again :weary:

Reply:
Sounds nice... Maybe it's a good thing that Microchip boards are not affordable...